Returning Blast datatypes to Galaxy core

peterjc opened the pull request:
This is a work-in-progress, I may re-write the branch history (again).

I am waiting to hear back from Edward Kirton (contacted by email) about the acknowledgment (_edit: and licence change_), I would welcome comments from @nsoranzo @bgruening as named contributors to this work.

Should I try to preserve any of the history from when this was worked on at https://github.com/peterjc/galaxy_blast/blob/master/datatypes/blast_datatypes/blast.py ?

My other major concern as asked on the mailing list https://lists.galaxyproject.org/pipermail/galaxy-dev/2016-August/024696.html is:

> Are there any pitfalls to worry about if the datatypes are
> already there with Galaxy and the tool shed version is
> installed on top? Or the tool shed version was installed
> but then Galaxy was updated to include the version
> bundled with Galaxy?

peterjc committed:
BLAST database *.loc.sample files

peterjc committed:
Include all three BLAST .loc files in data table conf

peterjc committed:
BLAST datatypes as of standalone blast_datatypes v0.0.21

Copied from separate BLAST+ Galaxy repository,
https://github.com/peterjc/galaxy_blast/blob/master/datatypes/blast_datatypes/blast.py

Currently available from the Galaxy Tool Shed,
http://toolshed.g2.bx.psu.edu/view/devteam/blast_datatypes

Primary contributors to the BLAST datatypes for Galaxy:

* Peter Cock (The James Hutton Institute, UK)
* Edward Kirton
* Nicola Soranzo
* Bjoern Gruening

This work and the associated BLAST+ wrappers can be cited as:

NCBI BLAST+ integrated into Galaxy.
P.J.A. Cock, J.M. Chilton, B. Gruening, J.E. Johnson, N. Soranzo
*GigaScience* 2015, 4:39
DOI: http://dx.doi.org/10.1186/s13742-015-0080-7

peterjc committed:
Remove unused imports

peterjc committed:
Missing import required for use as doctest

peterjc committed:
Sample blastxml file for sniff doctest

peterjc committed:
Add BLAST datatypes to datatypes_conf.xml.sample

peterjc committed:
Adjust import order as per Galaxy tox flake8 checks

peterjc committed:
More accurate module docstring

peterjc committed:
megablast_xml_parser_test1.blastxml should now sniff as blastxml

peterjc committed:
Ensure all BLAST datatype contributors in CONTRIBUTORS.md

peterjc committed:
Give my work email address as well.

My Galaxy work to date has been while employed at
The James Hutton Institute (UK).

bgruening commented:
@peterjc thanks for this! Afaik there are no issues in having datatypes from the TS and installed locally. Our production instance has this for snpeff for example and I never encountered any problems. I'm fine with leaving the commit history as they are, no extra work needed for me.

bgruening commented on lib/galaxy/datatypes/blast.py:
```diff
@@ -0,0 +1,281 @@
+
+
+class BlastXml(GenericXml):
+    """NCBI Blast XML Output data"""
```
I guess we should annotate them with the EDAM ontology as the other datatypes: https://bioportal.bioontology.org/ontologies/EDAM?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fformat_1333

@odoppelt can you give us a hand on this?

bgruening replied:
@peterjc can we add:

```
edam_format = "format_3331"
```

Not sure about `edam_data = "data_1381"` (http://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Fdata_1381)

It should look like this: https://github.com/galaxyproject/galaxy/blob/3fd0e405fd6cb3333397ec19e627a4cf6ae36fe6/lib/galaxy/datatypes/proteomics.py#L21

peterjc replied:
But is it http://edamontology.org/data_1381 _Alignment of exactly two molecular sequences_? In general a `blastxml` file contains multiple pairwise alignments from many query sequences and against multiple matched sequences. The parent entry http://edamontology.org/data_0863 _Alignment of multiple molecular sequences_ sounds safer to me.

Can you have multiple EDAM tags? http://edamontology.org/format_3331 _XML format as produced by the NCBI Blast package_ seems like the best choice if we can only pick one.

bgruening replied:
> But is it http://edamontology.org/data_1381 Alignment of exactly two molecular sequences? In general a blastxml file contains multiple pairwise alignments from many query sequences and against multiple matched sequences. The parent entry http://edamontology.org/data_0863 Alignment of multiple molecular sequences sounds safer to me.

I thought the alignments in a blastxml file are exactly of two sequences :) but we have multiple of these alignments.
You are probably correct.

> Can you have multiple EDAM tags? http://edamontology.org/format_3331 XML format as produced by the NCBI Blast package seems like the best choice if we can only pick one.

I think we can and should have only one. Go for the saver one if @odoppelt does not reply. She is our edam expert. Maybe Herve (@hmenager) can help as well?

peterjc replied:
Separately I want to add a new `blastxml2` datatype to Galaxy, and have filed https://github.com/edamontology/edamontology/issues/200 about adding an EDAM tag for the NCBI BLAST XML2 format.

martenson commented:
Do I understand this correctly that the blast datatypes were in distro first, then they moved to MTS and now they move back to distro?

peterjc commented:
@martenson Yes, really. Strange isn't it?

peterjc commented:
The failed "toolshed test" might be a false alarm,

```
Error Message

Installing blastxml_to_top_descr_0120 did not add new datatypes. Expected: 355. Found: 350
Stacktrace

Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/galaxy/test/shed_functional/functional/test_1140_simple_repository_dependency_multiple_owners.py", line 186, in test_0050_verify_repository_installation
    ( 'blastxml_to_top_descr_0120', expected_count, current_datatypes )
AssertionError: Installing blastxml_to_top_descr_0120 did not add new datatypes. Expected: 355. Found: 350
```

This would have tried to install the `blastxml` file format from the Tool Shed, which may have worked but because it and the other BLAST datatypes would have already been installed, it seems likely that the total number of datatypes wouldn't change.  (I'm making assumptions about how they are counted etc).

nsoranzo commented:
That's most probably the case, needs to be fixed somehow (change the datatype names in the tests?).

peterjc commented:
@nsoranzo the datatype names are deliberately the same. The fix would be making the test aware that the 5 new datatypes might be there already.

_Update:_ I'll take a look at https://github.com/galaxyproject/galaxy/blob/dev/test/shed_functional/functional/test_1140_simple_repository_dependency_multiple_owners.py#L185 tomorrow.

peterjc committed:
Tweak testing installing BLAST datatypes from Tool Shed

This is a short term pragmatic solution for the planned return
of the BLAST datatypes to the main Galaxy distribution. Long
term an additional test using an alternative Tool Shed defined
datatype may be needed.

peterjc commented:
I've not heard back from Edward, I may switch to @eskirton in the `CONTRIBUTORS.md` file in case he has any concerns about sharing his email address here.

Good news: The tests are all passing now :)

martenson commented:
I am 👍  on this, @peterjc please let me know when you want the WIP label removed.

peterjc commented:
Looking back at when this was removed from Galaxy and moved to the Tool Shed, it only covered the `blastxml` format, and that was in file `lib/galaxy/datatypes/xml.py`: https://github.com/galaxyproject/galaxy/commit/3d5fd7b5e27d33df02ea929f73250163f2e4e007

Is everyone happy with `lib/galaxy/datatypes/blast.py` defining the BLAST XML format and the BLAST databases?

bgruening commented on lib/galaxy/datatypes/blast.py:
```diff
@@ -0,0 +1,281 @@
+                    out.write(header)
+                    out.close()
+                    h.close()
+                    raise ValueError("BLAST XML file %s has too long a header!" % f)
```
Is the wording here correct?

peterjc replied:
I think it is grammatical, but perhaps `"The header in BLAST XML file %s is too long"` is better?

bgruening replied:
Sounds better for me.

bgruening commented:
@peterjc I'm fine with this! Would be nice to annotate it with the ontology before merging, if possible.

peterjc commented:
From going over the release notes and git logs, the direct contributors to the BLAST datatypes for Galaxy (and thus copyright holders) are:
- Peter Cock @peterjc
- Edward Kirton @eskirton
- Nicola Soranzo @nsoranzo 
- Bjoern Gruening @bgruening 

For the record, @bgruening and @nsoranzo are you both happy with adopting the _Academic Free License version 3.0_ here to match Galaxy?

@eskirton's original BLAST work is still on the Tool Shed, note there was no explicit license:

https://toolshed.g2.bx.psu.edu/view/edward-kirton/blast/

Looking back over old emails, although we didn't discuss the licensing, Edward gave permission for me to merge his BLAST database definitions into the datatype definitions I was looking after:

https://lists.galaxyproject.org/pipermail/galaxy-dev/2012-February/008544.html
https://lists.galaxyproject.org/pipermail/galaxy-dev/2012-April/009384.html

It was only at v0.0.16 of `blast_datatypes` (September 2013) that I explicitly put this under the _MIT license_:

https://toolshed.g2.bx.psu.edu/view/devteam/blast_datatypes

While his emails were an implicit permission to adopt the code and I treated it as such, I'd be much happier with Edward's explicit agreement to add his BLAST database datatype code to the Galaxy main repository under the _Academic Free License version 3.0_.

bgruening commented:
@peterjc permission granted :)

nsoranzo commented:
@peterjc Permission granted, of course.

> Is everyone happy with lib/galaxy/datatypes/blast.py defining the BLAST XML format and the BLAST databases?

I'm ambivalent on this, both `lib/galaxy/datatypes/xml.py` and `lib/galaxy/datatypes/blast.py` make sense for `blastxml` format, but classes inheriting from `GenericXml` are already present in `graph.py`, `molecules.py`, `proteomics.py` and `triples.py`, so probably it's good as it is now (and less work too!).

peterjc committed:
Clearer error message if BLAST XML header too long

bgruening commented:
I'm fine with merging as it is. @peterjc are you ready or do we want to wait for any response from @eskirton?

hmenager commented:
Hello,
Forgive my short reply, i'm out in the countryside and no good hardware and networking available ;)
As far as EDAM goes, format should be 3331 and data 0857 (sequence search results).
Cheers,
H.

bgruening commented:
@peterjc anything missing before we merge?
@hmenager any comments?

peterjc commented:
Based on the past email conversations, I doubt @eskirton would have any objections about re-licensing his contribution, but I've not heard back him so strictly speaking we could only merge this keeping the `blast.py` file under the liberal MIT license.

Hopefully the Galaxy project would be OK with that?

bgruening commented:
@galaxyproject/core are you ok with this license?
@peterjc can you merge dev into this branch please. Thanks!

jmchilton commented:
Other random little parts of Galaxy have different licenses - I think this is perfectly fine.

peterjc committed:
Explicitly label blast.py as MIT licensed.

Edward Kirton's contributions were made under the MIT license
while this code was developed outside of the main Galaxy
Project and distributed via the Galaxy Tool Shed). We have not
been able to contact him to authorise re-licensing under the
main Galaxy Project's license instead.

peterjc committed:
Mark BLAST XML with EDAM ontology terms.

As per Herve Menager, EDAM format should be 3331 and data 0857
(sequence search results).

peterjc commented:
OK, this has been rebased onto the current dev branch, `blast.py` explicitly labeled as MIT licensed, and I've hopefully added the EDAM ontology terms correctly (following the recommendation from Herve).

nsoranzo commented:
Thanks @peterjc, can you take a look at the failed Travis tests (py27-lint and py27-unit)?

peterjc commented:
@nsoranzo One trailing white space, and use of accents in Bjoern's name without an explicit encoding in the Python file - fixed and branch updated.

(@jmchilton Is there a pre-commit hook on the main Galaxy repository which I've neglected to turn on?)

hmenager commented:
all good for me, thank you @peterjc for the EDAM tagging ;)

bgruening commented:
Thanks a lot Peter!

peterjc commented:
Thanks everyone - we'll keep the Tool Shed version maintained for the medium term, as it will be some time before all the Galaxy instances world-wide have been updated to include the BLAST datatypes directly.

Tracking issue: https://github.com/peterjc/galaxy_blast/issues/75
